app/app.go

```diff
@@ -57,73 +72,72 @@ func Run(cfg conf.Config) {
 	// Set Peer default Push and Request handlers. These functions will handle
 	// request and push messages from all peers we connect to unless overridden
 	// for specific peers by calls like p.SetRequestHandler(someHandler)
-	peer.SetDefaultPushHandler(PushHandler)
-	peer.SetDefaultRequestHandler(RequestHandler)
+	a.PeerStore.SetDefaultPushHandler(a.PushHandler)
+	a.PeerStore.SetDefaultRequestHandler(a.RequestHandler)
 
 	// Start listening on the given interface and port so we can receive
 	// conenctions from other peers
 	log.Infof("Starting listener on %s:%d", cfg.Interface, cfg.Port)
-	peer.ListenAddr = fmt.Sprintf("%s:%d", cfg.Interface, cfg.Port)
+	a.PeerStore.ListenAddr = addr
 	go func() {
-		address := fmt.Sprintf("%s:%d", cfg.Interface, cfg.Port)
-		err := conn.Listen(address, peer.ConnectionHandler)
+		err := conn.Listen(addr, a.PeerStore.ConnectionHandler, wg)
 		if err != nil {
-			log.WithError(
-				err,
-			).Fatalf("Failed to listen on %s:%d", cfg.Interface, cfg.Port)
+			log.WithError(err).Fatalf("Failed to listen on %s", addr)
 		}
 	}()
 
 	// If the console flag was passed, redirect logs to a file and run the console
+	// NOTE: if the log file already exists we will exit with a fatal error here!
+	// This should stop people from running multiple Cumulus nodes that will try
+	// to log to the same file.
 	if cfg.Console {
-		logFile, err := os.OpenFile("logfile", os.O_WRONLY|os.O_CREATE, 0755)
+		logFile, err := os.OpenFile("logfile", os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
 		if err != nil {
-			log.WithError(err).Fatal("Failed to redirect logs to log file")
+			log.WithError(err).Fatal("Failed to redirect logs to file")
 		}
-		log.Warn("Redirecting logs to logfile")
+		log.Warn("Redirecting logs to file")
 		log.SetOutput(logFile)
-		go RunConsole()
+		wg.Add(1)
+		go RunConsole(a.PeerStore, wg)
 	}
 
 	if len(config.Target) > 0 {
 		// Connect to the target and discover its peers.
-		ConnectAndDiscover(cfg.Target)
+		a.ConnectAndDiscover(cfg.Target)
 	}
 
 	// Try maintain as close to peer.MaxPeers connections as possible while this
 	// peer is running
-	go peer.MaintainConnections()
+	go a.PeerStore.MaintainConnections(wg)
 
 	// Request the blockchain.
 	if chain == nil {
 		log.Info("Request blockchain from peers not yet implemented.")
 		initializeChain()
 	}
 
-	// Return to command line.
-	select {} // Hang main thread. Everything happens in goroutines from here
+	// Wait for goroutines to start
+	wg.Wait()
 }
 
 // ConnectAndDiscover tries to connect to a target and discover its peers.
-func ConnectAndDiscover(target string) {
+func (a App) ConnectAndDiscover(target string) {
```

Comment:
Let's have the received be an app pointer to avoid copying. Not a big deal but standard practice